packages/next/src/client/components/app-router.tsx

```diff
@@ -592,8 +603,46 @@ export default function AppRouter({
   globalErrorComponent: ErrorComponent
   assetPrefix: string
 }) {
+  if (process.env.__NEXT_APP_NAV_FAIL_HANDLING) {
```

Comment:
Probably would be worth throwing this into a custom hook just so we can compartmentalize the experimental stuff from the main `app-router` render code